api/src/controllers/changes.js

```diff
@@ -1,578 +1,396 @@
-    }
-    // use setTimeout to break recursion so stack doesn't blow out
-    setTimeout(function() {
-      init(since);
-    }, 1000);
-  });
 };
 
-module.exports = {
-  request: function(proxy, req, res) {
-    if (!inited) {
-      init();
-    }
-    auth.getUserCtx(req, function(err, userCtx) {
-      if (err) {
-        return serverUtils.error(err, req, res);
-      }
+const request = (proxy, req, res) => {
+  init();
 
-      if (req.query.feed === 'longpoll' ||
-          req.query.feed === 'continuous' ||
-          req.query.feed === 'eventsource') {
+  return auth
+    .getUserCtx(req)
+    .then(userCtx => {
+      if (isLongpoll(req)) {
         // Disable nginx proxy buffering to allow hearbeats for long-running feeds
         // Issue: #2363
         res.setHeader('X-Accel-Buffering', 'no');
       }
 
       if (auth.isAdmin(userCtx)) {
-        proxy.web(req, res);
-      } else {
-        const feed = {
-          id: uuid(),
-          req: req,
-          res: res,
-          userCtx: userCtx,
-          upstreamBatches: {},
-        };
-        const logPrefix = `[feed:${feed.id}]`;
-        feed.log = (...message) => console.log(logPrefix, ...message);
-
-        req.on('close', function() {
-          const uBatchCount = Object.keys(feed.upstreamBatches).length;
-          feed.log(`[event:close] outstanding ${uBatchCount} upstream batches will be cancelled, and feed will be cleaned up.`);
-          abortUpstreamBatches(feed, 'response-closed', 'WARN unresolved upstream batch(es) after response was closed:');
-          cleanUp(feed);
-        });
-        initFeed(feed, function(err) {
-          if (err) {
-            return serverUtils.error(err, req, res);
-          }
-          if (req.query.feed === 'longpoll') {
-            // watch for newly added docs
-            continuousFeeds.push(feed);
-          }
-          res.type('json');
-          defibrillator(feed);
-          feed.log('initFeed().callback', 'calling getChanges()…');
-          getChanges(feed);
-        });
+        return proxy.web(req, res);
       }
-    });
-  },
-  _getReplicationKey: getReplicationKey // used for testing
+      res.type('json');
+      return processRequest(req, res, userCtx);
+    })
+    .catch(() => serverUtils.notLoggedIn(req, res));
 };
 
-function startTimer() {
-  return Date.now();
-}
-
-function endTimer(name, start) {
-  var diff = Date.now() - start;
-  console.log('TIMED SECTION COMPLETE', name, diff, 'ms');
-}
+module.exports = {
+  request: request,
+};
 
 // used for testing
 if (process.env.UNIT_TEST_ENV) {
   _.extend(module.exports, {
-    _reset: function() {
-      continuousFeeds = [];
+    _init: init,
+    _initFeed: initFeed,
+    _processChange: processChange,
+    _writeDownstream: writeDownstream,
+    _processPendingChanges: processPendingChanges,
+    _appendChange: appendChange,
+    _mergeResults: mergeResults,
+    _tombstoneUtils: tombstoneUtils,
+    _: _,
```

Comment:
I assume this is here so we can mock `_` calls? Generally _ calls should be tested as part of the unit to ensure they're being used properly - essentially I consider them language extensions and we don't mock language features like `forEach`.

Reply:
The only reason `_` was mocked was because I was struggling making it's `debounce` function follow sinon's `fakeTimers`. In the end, exporting `_` wasn't needed, but mocking `_.now` was.